File: data_mcp/gpt-5/scala/rep7/src/Main.scala
//> using scala "3.4.2"
//> using dep "com.softwaremill.sttp.tapir::tapir-http4s-server:1.10.7"
//> using dep "org.http4s::http4s-ember-server:0.23.26"
//> using dep "org.http4s::http4s-circe:0.23.26"
//> using dep "io.circe::circe-generic:0.14.7"
//> using dep "io.circe::circe-parser:0.14.7"
//> using dep "ch.qos.logback:logback-classic:1.5.6"

import cats.effect._
import cats.syntax.all._
import org.http4s._
import org.http4s.dsl.io._
import org.http4s.ember.server.EmberServerBuilder
import org.http4s.implicits._
import org.http4s.circe._
import io.circe._
import io.circe.generic.semiauto._
import io.circe.syntax._
import java.time._
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import scala.jdk.CollectionConverters._
import com.comcast.ip4s._

object Main extends IOApp.Simple:

  case class User(id: Int, username: String)
  case class UserWithPassword(user: User, password: String)

  case class Todo(
      id: Int,
      userId: Int,
      title: String,
      description: String,
      completed: Boolean,
      created_at: String,
      updated_at: String
  )

  given Encoder[User] = deriveEncoder
  given Decoder[User] = deriveDecoder

  given Encoder[Todo] = deriveEncoder
  given Decoder[Todo] = deriveDecoder

  case class RegisterReq(username: String, password: String)
  case class LoginReq(username: String, password: String)
  case class ChangePasswordReq(old_password: String, new_password: String)
  case class CreateTodoReq(title: String, description: Option[String])
  case class UpdateTodoReq(title: Option[String], description: Option[String], completed: Option[Boolean])

  given Decoder[RegisterReq] = deriveDecoder
  given Decoder[LoginReq] = deriveDecoder
  given Decoder[ChangePasswordReq] = deriveDecoder
  given Decoder[CreateTodoReq] = deriveDecoder
  given Decoder[UpdateTodoReq] = deriveDecoder

  private val contentJson: Header.Raw = Header.Raw(ci"Content-Type", "application/json")

  def jsonResponse(status: Status, json: Json): Response[IO] =
    Response[IO](status).withHeaders(Headers(contentJson)).withEntity(json.noSpaces)

  def jsonError(status: Status, message: String): Response[IO] =
    jsonResponse(status, Json.obj("error" -> Json.fromString(message)))

  object State:
    private val usersById = new ConcurrentHashMap[Int, UserWithPassword]()
    private val usersByName = new ConcurrentHashMap[String, Int]()
    private val sessions = new ConcurrentHashMap[String, Int]() // token -> userId
    private val todosById = new ConcurrentHashMap[Int, Todo]()
    private var userSeq: Int = 0
    private var todoSeq: Int = 0

    private val usernameRegex = "^[a-zA-Z0-9_]{3,50}$".r

    private def nextUserId(): Int = synchronized { userSeq += 1; userSeq }
    private def nextTodoId(): Int = synchronized { todoSeq += 1; todoSeq }

    def register(username: String, password: String): Either[Response[IO], User] =
      if username == null || password == null then Left(jsonError(Status.BadRequest, "Invalid request"))
      else if !usernameRegex.pattern.matcher(username).matches() then Left(jsonError(Status.BadRequest, "Invalid username"))
      else if password.length < 8 then Left(jsonError(Status.BadRequest, "Password too short"))
      else synchronized {
        if usersByName.containsKey(username) then Left(jsonError(Status.Conflict, "Username already exists"))
        else
          val id = nextUserId()
          val user = User(id, username)
          usersById.put(id, UserWithPassword(user, password))
          usersByName.put(username, id)
          Right(user)
      }

    def login(username: String, password: String): Either[Response[IO], (User, String)] =
      val uid = Option(usersByName.get(username)).getOrElse(-1)
      val maybe = Option(usersById.get(uid))
      maybe match
        case Some(uwp) if uwp.password == password =>
          val token = UUID.randomUUID().toString.replaceAll("-", "")
          sessions.put(token, uwp.user.id)
          Right((uwp.user, token))
        case _ => Left(jsonError(Status.Unauthorized, "Invalid credentials"))

    def userBySession(token: String): Option[User] =
      Option(sessions.get(token)).flatMap(id => Option(usersById.get(id)).map(_.user))

    def invalidateSession(token: String): Unit = sessions.remove(token)

    def changePassword(userId: Int, oldPassword: String, newPassword: String): Either[Response[IO], Unit] =
      val maybe = Option(usersById.get(userId))
      maybe match
        case Some(uwp) if uwp.password == oldPassword =>
          if newPassword.length < 8 then Left(jsonError(Status.BadRequest, "Password too short"))
          else
            usersById.put(userId, uwp.copy(password = newPassword))
            Right(())
        case _ => Left(jsonError(Status.Unauthorized, "Invalid credentials"))

    private def nowIso(): String =
      java.time.Instant.now().truncatedTo(java.time.temporal.ChronoUnit.SECONDS).toString

    def listTodos(userId: Int): List[Todo] =
      todosById.values().asScala.toList.filter(_.userId == userId).sortBy(_.id)

    def createTodo(userId: Int, title: String, description: Option[String]): Either[Response[IO], Todo] =
      if title == null || title.trim.isEmpty then Left(jsonError(Status.BadRequest, "Title is required"))
      else synchronized {
        val id = nextTodoId()
        val ts = nowIso()
        val todo = Todo(id, userId, title, description.getOrElse(""), false, ts, ts)
        todosById.put(id, todo)
        Right(todo)
      }

    def getTodo(userId: Int, id: Int): Either[Response[IO], Todo] =
      Option(todosById.get(id)) match
        case Some(t) if t.userId == userId => Right(t)
        case _ => Left(jsonError(Status.NotFound, "Todo not found"))

    def updateTodo(userId: Int, id: Int, patch: UpdateTodoReq): Either[Response[IO], Todo] =
      Option(todosById.get(id)) match
        case Some(t) if t.userId == userId =>
          patch.title match
            case Some(v) if v.trim.isEmpty => Left(jsonError(Status.BadRequest, "Title is required"))
            case _ =>
              val newTitle = patch.title.getOrElse(t.title)
              val newDesc = patch.description.getOrElse(t.description)
              val newCompleted = patch.completed.getOrElse(t.completed)
              val updated = t.copy(title = newTitle, description = newDesc, completed = newCompleted, updated_at = nowIso())
              todosById.put(id, updated)
              Right(updated)
        case _ => Left(jsonError(Status.NotFound, "Todo not found"))

    def deleteTodo(userId: Int, id: Int): Either[Response[IO], Unit] =
      Option(todosById.get(id)) match
        case Some(t) if t.userId == userId =>
          todosById.remove(id)
          Right(())
        case _ => Left(jsonError(Status.NotFound, "Todo not found"))

  end State

  object Auth:
    private val CookieName = "session_id"

    def extractSession(req: Request[IO]): Option[String] =
      req.cookies.find(_.name == CookieName).map(_.content)

    def requireAuth(req: Request[IO]): Either[Response[IO], User] =
      extractSession(req) match
        case Some(token) => State.userBySession(token).toRight(jsonError(Status.Unauthorized, "Authentication required"))
        case None        => Left(jsonError(Status.Unauthorized, "Authentication required"))

    def setCookie(token: String): Header.Raw =
      Header.Raw(ci"Set-Cookie", s"$CookieName=$token; Path=/; HttpOnly")

    def invalidate(req: Request[IO]): Unit =
      extractSession(req).foreach(State.invalidateSession)

  end Auth

  given EntityDecoder[IO, Json] = jsonOf[IO, Json]

  private def parseJsonBody[T: Decoder](req: Request[IO]): IO[Either[Response[IO], T]] =
    req.attemptAs[Json].value.map {
      case Left(_) => Left(jsonError(Status.BadRequest, "Invalid JSON"))
      case Right(json) =>
        json.as[T] match
          case Left(_) => Left(jsonError(Status.BadRequest, "Invalid request"))
          case Right(value) => Right(value)
    }

  private def userJson(u: User): Json = Json.obj("id" -> Json.fromInt(u.id), "username" -> Json.fromString(u.username))

  val routes: HttpRoutes[IO] = HttpRoutes.of[IO] {
    // POST /register
    case req @ POST -> Root / "register" =>
      parseJsonBody[RegisterReq](req).flatMap {
        case Left(err) => IO.pure(err)
        case Right(RegisterReq(username, password)) =>
          State.register(username, password) match
            case Left(err) => IO.pure(err)
            case Right(user) => IO.pure(jsonResponse(Status.Created, userJson(user)))
      }

    // POST /login
    case req @ POST -> Root / "login" =>
      parseJsonBody[LoginReq](req).flatMap {
        case Left(err) => IO.pure(err)
        case Right(LoginReq(username, password)) =>
          State.login(username, password) match
            case Left(err) => IO.pure(err)
            case Right((user, token)) =>
              val resp = jsonResponse(Status.Ok, userJson(user))
              IO.pure(resp.putHeaders(Auth.setCookie(token)))
      }

    // POST /logout
    case req @ POST -> Root / "logout" =>
      Auth.requireAuth(req) match
        case Left(err) => IO.pure(err)
        case Right(_) =>
          Auth.invalidate(req)
          IO.pure(jsonResponse(Status.Ok, Json.obj()))

    // GET /me
    case req @ GET -> Root / "me" =>
      Auth.requireAuth(req) match
        case Left(err) => IO.pure(err)
        case Right(user) => IO.pure(jsonResponse(Status.Ok, userJson(user)))

    // PUT /password
    case req @ PUT -> Root / "password" =>
      Auth.requireAuth(req) match
        case Left(err) => IO.pure(err)
        case Right(user) =>
          parseJsonBody[ChangePasswordReq](req).flatMap {
            case Left(err) => IO.pure(err)
            case Right(ChangePasswordReq(oldp, newp)) =>
              State.changePassword(user.id, oldp, newp) match
                case Left(err) => IO.pure(err)
                case Right(_) => IO.pure(jsonResponse(Status.Ok, Json.obj()))
          }

    // GET /todos
    case req @ GET -> Root / "todos" =>
      Auth.requireAuth(req) match
        case Left(err) => IO.pure(err)
        case Right(user) =>
          val list = State.listTodos(user.id).map { t =>
            Json.obj(
              "id" -> Json.fromInt(t.id),
              "title" -> Json.fromString(t.title),
              "description" -> Json.fromString(t.description),
              "completed" -> Json.fromBoolean(t.completed),
              "created_at" -> Json.fromString(t.created_at),
              "updated_at" -> Json.fromString(t.updated_at)
            )
          }
          IO.pure(jsonResponse(Status.Ok, Json.fromValues(list)))

    // POST /todos
    case req @ POST -> Root / "todos" =>
      Auth.requireAuth(req) match
        case Left(err) => IO.pure(err)
        case Right(user) =>
          parseJsonBody[CreateTodoReq](req).flatMap {
            case Left(err) => IO.pure(err)
            case Right(CreateTodoReq(title, desc)) =>
              State.createTodo(user.id, title, desc) match
                case Left(err) => IO.pure(err)
                case Right(t) => IO.pure(jsonResponse(Status.Created, todoJsonPublic(t)))
          }

    // GET /todos/:id
    case req @ GET -> Root / "todos" / IntVar(id) =>
      Auth.requireAuth(req) match
        case Left(err) => IO.pure(err)
        case Right(user) =>
          State.getTodo(user.id, id) match
            case Left(err) => IO.pure(err)
            case Right(t) => IO.pure(jsonResponse(Status.Ok, todoJsonPublic(t)))

    // PUT /todos/:id
    case req @ PUT -> Root / "todos" / IntVar(id) =>
      Auth.requireAuth(req) match
        case Left(err) => IO.pure(err)
        case Right(user) =>
          parseJsonBody[UpdateTodoReq](req).flatMap {
            case Left(err) => IO.pure(err)
            case Right(patch) =>
              State.updateTodo(user.id, id, patch) match
                case Left(err) => IO.pure(err)
                case Right(t) => IO.pure(jsonResponse(Status.Ok, todoJsonPublic(t)))
          }

    // DELETE /todos/:id
    case req @ DELETE -> Root / "todos" / IntVar(id) =>
      Auth.requireAuth(req) match
        case Left(err) => IO.pure(err)
        case Right(user) =>
          State.deleteTodo(user.id, id) match
            case Left(err) => IO.pure(err)
            case Right(_) =>
              // 204 with no body and no Content-Type
              IO.pure(Response[IO](Status.NoContent))
  }

  private def todoJsonPublic(t: Todo): Json =
    Json.obj(
      "id" -> Json.fromInt(t.id),
      "title" -> Json.fromString(t.title),
      "description" -> Json.fromString(t.description),
      "completed" -> Json.fromBoolean(t.completed),
      "created_at" -> Json.fromString(t.created_at),
      "updated_at" -> Json.fromString(t.updated_at)
    )

  def app: HttpApp[IO] =
    val httpApp = routes.orNotFound
    Kleisli { (req: Request[IO]) =>
      httpApp(req).map { resp =>
        if resp.status == Status.NoContent then resp
        else
          // Ensure Content-Type: application/json on all non-DELETE responses
          val hasCt = resp.headers.get[headers.`Content-Type`].nonEmpty
          if hasCt then resp
          else resp.withHeaders(resp.headers ++ Headers(contentJson))
      }
    }

  def run: IO[Unit] =
    for
      args <- IO(sys.args.toList)
      port =
        args match
          case "--port" :: p :: _ => p.toIntOption.getOrElse(8080)
          case _ => sys.env.get("PORT").flatMap(_.toIntOption).getOrElse(8080)
      _ <- EmberServerBuilder.default[IO]
        .withHost(ipv4"0.0.0.0")
        .withPort(Port.fromInt(port).get)
        .withHttpApp(app)
        .build
        .useForever
    yield ()

end Main

File: data_mcp/gpt-5/java/rep7/Main.java
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;
import com.google.gson.*;
import com.google.gson.annotations.SerializedName;

import java.io.*;
import java.net.InetSocketAddress;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

public class Main {
    private static final int DEFAULT_PORT = 8080;

    public static void main(String[] args) throws Exception {
        int port = DEFAULT_PORT;
        for (int i = 0; i < args.length; i++) {
            if ("--port".equals(args[i]) && i + 1 < args.length) {
                try {
                    port = Integer.parseInt(args[i + 1]);
                    i++;
                } catch (NumberFormatException e) {
                    System.err.println("Invalid port number");
                    System.exit(1);
                }
            }
        }

        InMemoryStore store = new InMemoryStore();
        Server server = new Server(port, store);
        server.start();
    }
}

class Server {
    private final int port;
    private final InMemoryStore store;
    private HttpServer httpServer;

    private static final Gson gson = new GsonBuilder().serializeNulls().create();

    public Server(int port, InMemoryStore store) {
        this.port = port;
        this.store = store;
    }

    public void start() throws IOException {
        httpServer = HttpServer.create(new InetSocketAddress("0.0.0.0", port), 0);
        httpServer.createContext("/", new RootHandler(store));
        httpServer.setExecutor(Executors.newCachedThreadPool());
        httpServer.start();
        System.out.println("Server started on 0.0.0.0:" + port);
    }

    static class RootHandler implements HttpHandler {
        private final InMemoryStore store;
        private final DateTimeFormatter isoFormatter = DateTimeFormatter.ISO_INSTANT.withZone(ZoneOffset.UTC);

        RootHandler(InMemoryStore store) {
            this.store = store;
        }

        @Override
        public void handle(HttpExchange exchange) throws IOException {
            try {
                String method = exchange.getRequestMethod();
                URI uri = exchange.getRequestURI();
                String path = uri.getPath();

                if (path.equals("/register") && method.equals("POST")) {
                    handleRegister(exchange);
                    return;
                }
                if (path.equals("/login") && method.equals("POST")) {
                    handleLogin(exchange);
                    return;
                }
                if (path.equals("/logout") && method.equals("POST")) {
                    User user = authenticate(exchange);
                    if (user == null) return; // authenticate already sent 401
                    handleLogout(exchange);
                    return;
                }
                if (path.equals("/me") && method.equals("GET")) {
                    User user = authenticate(exchange);
                    if (user == null) return;
                    handleMe(exchange, user);
                    return;
                }
                if (path.equals("/password") && method.equals("PUT")) {
                    User user = authenticate(exchange);
                    if (user == null) return;
                    handlePasswordChange(exchange, user);
                    return;
                }
                if (path.equals("/todos")) {
                    User user = authenticate(exchange);
                    if (user == null) return;
                    if (method.equals("GET")) {
                        handleTodosList(exchange, user);
                        return;
                    } else if (method.equals("POST")) {
                        handleTodosCreate(exchange, user);
                        return;
                    }
                }
                if (path.startsWith("/todos/") && path.length() > "/todos/".length()) {
                    User user = authenticate(exchange);
                    if (user == null) return;
                    String idStr = path.substring("/todos/".length());
                    Integer id = parseId(idStr);
                    if (id == null) {
                        sendJson(exchange, 404, error("Todo not found"));
                        return;
                    }
                    if (method.equals("GET")) {
                        handleTodoGet(exchange, user, id);
                        return;
                    } else if (method.equals("PUT")) {
                        handleTodoUpdate(exchange, user, id);
                        return;
                    } else if (method.equals("DELETE")) {
                        handleTodoDelete(exchange, user, id);
                        return;
                    }
                }

                // Not found or method not allowed
                sendJson(exchange, 404, error("Not found"));
            } catch (Exception e) {
                e.printStackTrace();
                try {
                    sendJson(exchange, 500, error("Internal server error"));
                } catch (Exception ignore) {}
            } finally {
                try { exchange.close(); } catch (Exception ignore) {}
            }
        }

        private void handleRegister(HttpExchange exchange) throws IOException {
            String body = readBody(exchange);
            if (body == null) body = "";
            JsonObject obj;
            try {
                obj = JsonParser.parseString(body).getAsJsonObject();
            } catch (Exception e) {
                sendJson(exchange, 400, error("Invalid JSON"));
                return;
            }
            String username = getAsString(obj, "username");
            String password = getAsString(obj, "password");

            if (username == null || !username.matches("^[A-Za-z0-9_]{3,50}$")) {
                sendJson(exchange, 400, error("Invalid username"));
                return;
            }
            if (password == null || password.length() < 8) {
                sendJson(exchange, 400, error("Password too short"));
                return;
            }
            synchronized (store) {
                if (store.usernameToUser.containsKey(username)) {
                    sendJson(exchange, 409, error("Username already exists"));
                    return;
                }
                User user = store.createUser(username, password);
                JsonObject resp = new JsonObject();
                resp.addProperty("id", user.id);
                resp.addProperty("username", user.username);
                sendJson(exchange, 201, resp);
            }
        }

        private void handleLogin(HttpExchange exchange) throws IOException {
            String body = readBody(exchange);
            if (body == null) body = "";
            JsonObject obj;
            try {
                obj = JsonParser.parseString(body).getAsJsonObject();
            } catch (Exception e) {
                sendJson(exchange, 400, error("Invalid JSON"));
                return;
            }
            String username = getAsString(obj, "username");
            String password = getAsString(obj, "password");

            if (username == null || password == null) {
                sendJson(exchange, 401, error("Invalid credentials"));
                return;
            }
            User user = store.usernameToUser.get(username);
            if (user == null || !user.password.equals(password)) {
                sendJson(exchange, 401, error("Invalid credentials"));
                return;
            }
            String token = UUID.randomUUID().toString().replaceAll("-", "");
            store.sessions.put(token, user.id);
            exchange.getResponseHeaders().add("Set-Cookie", "session_id=" + token + "; Path=/; HttpOnly");
            JsonObject resp = new JsonObject();
            resp.addProperty("id", user.id);
            resp.addProperty("username", user.username);
            sendJson(exchange, 200, resp);
        }

        private void handleLogout(HttpExchange exchange) throws IOException {
            String token = getSessionToken(exchange);
            if (token != null) {
                store.sessions.remove(token);
            }
            JsonObject resp = new JsonObject();
            sendJson(exchange, 200, resp);
        }

        private void handleMe(HttpExchange exchange, User user) throws IOException {
            JsonObject resp = new JsonObject();
            resp.addProperty("id", user.id);
            resp.addProperty("username", user.username);
            sendJson(exchange, 200, resp);
        }

        private void handlePasswordChange(HttpExchange exchange, User user) throws IOException {
            String body = readBody(exchange);
            if (body == null) body = "";
            JsonObject obj;
            try {
                obj = JsonParser.parseString(body).getAsJsonObject();
            } catch (Exception e) {
                sendJson(exchange, 400, error("Invalid JSON"));
                return;
            }
            String oldPassword = getAsString(obj, "old_password");
            String newPassword = getAsString(obj, "new_password");

            if (oldPassword == null || !user.password.equals(oldPassword)) {
                sendJson(exchange, 401, error("Invalid credentials"));
                return;
            }
            if (newPassword == null || newPassword.length() < 8) {
                sendJson(exchange, 400, error("Password too short"));
                return;
            }
            user.password = newPassword;
            JsonObject resp = new JsonObject();
            sendJson(exchange, 200, resp);
        }

        private void handleTodosList(HttpExchange exchange, User user) throws IOException {
            List<Todo> list = new ArrayList<>();
            for (Todo t : store.todos.values()) {
                if (t.userId == user.id) list.add(t);
            }
            list.sort(Comparator.comparingInt(a -> a.id));
            JsonArray arr = new JsonArray();
            for (Todo t : list) {
                arr.add(todoToJson(t));
            }
            sendJson(exchange, 200, arr);
        }

        private void handleTodosCreate(HttpExchange exchange, User user) throws IOException {
            String body = readBody(exchange);
            if (body == null) body = "";
            JsonObject obj;
            try {
                obj = JsonParser.parseString(body).getAsJsonObject();
            } catch (Exception e) {
                sendJson(exchange, 400, error("Invalid JSON"));
                return;
            }
            String title = getAsString(obj, "title");
            String description = getAsString(obj, "description");
            if (title == null || title.trim().isEmpty()) {
                sendJson(exchange, 400, error("Title is required"));
                return;
            }
            if (description == null) description = "";
            Instant now = Instant.now().truncatedTo(ChronoUnit.SECONDS);
            Todo todo = store.createTodo(user.id, title, description, now);
            sendJson(exchange, 201, todoToJson(todo));
        }

        private void handleTodoGet(HttpExchange exchange, User user, int id) throws IOException {
            Todo t = store.todos.get(id);
            if (t == null || t.userId != user.id) {
                sendJson(exchange, 404, error("Todo not found"));
                return;
            }
            sendJson(exchange, 200, todoToJson(t));
        }

        private void handleTodoUpdate(HttpExchange exchange, User user, int id) throws IOException {
            Todo t = store.todos.get(id);
            if (t == null || t.userId != user.id) {
                sendJson(exchange, 404, error("Todo not found"));
                return;
            }
            String body = readBody(exchange);
            if (body == null) body = "";
            JsonObject obj;
            try {
                obj = JsonParser.parseString(body).getAsJsonObject();
            } catch (Exception e) {
                sendJson(exchange, 400, error("Invalid JSON"));
                return;
            }
            boolean modified = false;
            if (obj.has("title")) {
                JsonElement el = obj.get("title");
                if (el.isJsonNull()) {
                    // ignore nulls, but treat as empty? Spec does not allow null - we'll treat null as empty invalid
                    sendJson(exchange, 400, error("Title is required"));
                    return;
                }
                String title = el.getAsString();
                if (title.trim().isEmpty()) {
                    sendJson(exchange, 400, error("Title is required"));
                    return;
                }
                t.title = title;
                modified = true;
            }
            if (obj.has("description")) {
                if (obj.get("description").isJsonNull()) {
                    t.description = "";
                } else {
                    t.description = obj.get("description").getAsString();
                }
                modified = true;
            }
            if (obj.has("completed")) {
                if (obj.get("completed").isJsonNull()) {
                    // ignore null - do not modify
                } else {
                    t.completed = obj.get("completed").getAsBoolean();
                    modified = true;
                }
            }
            if (modified) {
                t.updatedAt = Instant.now().truncatedTo(ChronoUnit.SECONDS);
            }
            sendJson(exchange, 200, todoToJson(t));
        }

        private void handleTodoDelete(HttpExchange exchange, User user, int id) throws IOException {
            Todo t = store.todos.get(id);
            if (t == null || t.userId != user.id) {
                sendJson(exchange, 404, error("Todo not found"));
                return;
            }
            store.todos.remove(id);
            // 204 No Content, no body
            sendNoContent(exchange);
        }

        private User authenticate(HttpExchange exchange) throws IOException {
            String token = getSessionToken(exchange);
            if (token == null) {
                sendJson(exchange, 401, error("Authentication required"));
                return null;
            }
            Integer userId = store.sessions.get(token);
            if (userId == null) {
                sendJson(exchange, 401, error("Authentication required"));
                return null;
            }
            User user = store.users.get(userId);
            if (user == null) {
                sendJson(exchange, 401, error("Authentication required"));
                return null;
            }
            return user;
        }

        private String getSessionToken(HttpExchange exchange) {
            List<String> cookies = exchange.getRequestHeaders().get("Cookie");
            if (cookies == null) return null;
            for (String header : cookies) {
                String[] parts = header.split(";\\s*");
                for (String part : parts) {
                    int idx = part.indexOf('=');
                    if (idx > 0) {
                        String name = part.substring(0, idx).trim();
                        String value = part.substring(idx + 1).trim();
                        if (name.equals("session_id")) {
                            return value;
                        }
                    }
                }
            }
            return null;
        }

        private Integer parseId(String s) {
            try {
                return Integer.parseInt(s);
            } catch (Exception e) {
                return null;
            }
        }

        private String formatInstant(Instant inst) {
            return DateTimeFormatter.ISO_INSTANT.format(inst);
        }

        private JsonObject todoToJson(Todo t) {
            JsonObject obj = new JsonObject();
            obj.addProperty("id", t.id);
            obj.addProperty("title", t.title);
            obj.addProperty("description", t.description);
            obj.addProperty("completed", t.completed);
            obj.addProperty("created_at", formatInstant(t.createdAt));
            obj.addProperty("updated_at", formatInstant(t.updatedAt));
            return obj;
        }

        private void sendJson(HttpExchange exchange, int status, JsonElement json) throws IOException {
            byte[] bytes = json.toString().getBytes(StandardCharsets.UTF_8);
            exchange.getResponseHeaders().set("Content-Type", "application/json");
            exchange.sendResponseHeaders(status, bytes.length);
            try (OutputStream os = exchange.getResponseBody()) {
                os.write(bytes);
            }
        }

        private void sendNoContent(HttpExchange exchange) throws IOException {
            // No Content with no body
            exchange.sendResponseHeaders(204, -1);
        }

        private String readBody(HttpExchange exchange) throws IOException {
            try (InputStream is = exchange.getRequestBody()) {
                ByteArrayOutputStream baos = new ByteArrayOutputStream();
                byte[] buffer = new byte[4096];
                int r;
                while ((r = is.read(buffer)) != -1) {
                    baos.write(buffer, 0, r);
                }
                return baos.toString(StandardCharsets.UTF_8);
            }
        }

        private String getAsString(JsonObject obj, String member) {
            if (obj == null || !obj.has(member) || obj.get(member).isJsonNull()) return null;
            try { return obj.get(member).getAsString(); } catch (Exception e) { return null; }
        }

        private JsonObject error(String message) {
            JsonObject obj = new JsonObject();
            obj.addProperty("error", message);
            return obj;
        }
    }
}

class InMemoryStore {
    final Map<Integer, User> users = new ConcurrentHashMap<>();
    final Map<String, User> usernameToUser = new ConcurrentHashMap<>();
    final Map<String, Integer> sessions = new ConcurrentHashMap<>();
    final Map<Integer, Todo> todos = new ConcurrentHashMap<>();

    private final AtomicInteger userIdSeq = new AtomicInteger(1);
    private final AtomicInteger todoIdSeq = new AtomicInteger(1);

    synchronized User createUser(String username, String password) {
        int id = userIdSeq.getAndIncrement();
        User user = new User(id, username, password);
        users.put(id, user);
        usernameToUser.put(username, user);
        return user;
    }

    synchronized Todo createTodo(int userId, String title, String description, Instant now) {
        int id = todoIdSeq.getAndIncrement();
        Todo t = new Todo(id, userId, title, description, false, now, now);
        todos.put(id, t);
        return t;
    }
}

class User {
    final int id;
    final String username;
    String password;

    User(int id, String username, String password) {
        this.id = id;
        this.username = username;
        this.password = password;
    }
}

class Todo {
    final int id;
    final int userId;
    String title;
    String description;
    boolean completed;
    Instant createdAt;
    Instant updatedAt;

    Todo(int id, int userId, String title, String description, boolean completed, Instant createdAt, Instant updatedAt) {
        this.id = id;
        this.userId = userId;
        this.title = title;
        this.description = description;
        this.completed = completed;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }
}
